### PR TITLE
Implement timed effects for Hue lights

### PR DESCRIPTION
**Pull request by duvholt**
Using timed_effects made it much easier to inspect the zigbee message output of running the sunrise effect, so I tweaked EffectDuration to be as accurate as possible. Unfortunately that revealed some edge cases that didn't fit the existing model which made the conversion function uglier.

To test "timed effects" send a PUT request with 
```json
{
  "on": {
    "on": true
  },
  "timed_effects": {
    "effect": "sunrise",
    "duration": 60000
  }
}
```

Duration is in milliseconds. 

One thing that is missing is updating `timed_effects.status`. Since that can easily lead to concurrency issues I'll keep that out of the scope for now.

Oh, and as a bonus I added the sunset effect 🌆 
```json
{
  "on": {
    "on": true
  },
  "timed_effects": {
    "effect": "sunset",
    "duration": 60000
  }
}
```

### CODE REVIEW

**Comment by chrivers** [comment]
Hey - sorry I completely overlooked this somehow :sweat_smile: 

I'll take a look now!

**Comment by chrivers** [comment]
Do we know what `timed_effects.status` is supposed to contain? What's the expected value(s)?

File: crates/hue/src/api/light.rs [hunk 1]
```diff
@@ -501,7 +501,6 @@ pub enum LightEffect {
     Cosmos,
     Sunbeam,
     Enchant,
-    Sunrise,
```
Comment (chrivers) [inline crates/hue/src/api/light.rs:504]: I guess it doesn't cause any problems to remove this effect?

Was it supposed to be a timed effect the whole time?

Comment (duvholt) [reply crates/hue/src/api/light.rs:504]: Yes, I don't think it makes sense to use sunrise/sunset as an LightEffect as it doesn't repeat like the other "LightEffect"s. 
I added it there before I knew about the timed effect field.

File: crates/hue/src/effect_duration.rs [hunk 2]
```diff
@@ -91,4 +106,91 @@ mod tests {
...
+
+        for (input_ms, zigbee_data) in input {
+            let nearest_second: u32 = (f64::from(input_ms) / 1000.0).round() as u32;
+            let ed = EffectDuration::from_ms(input_ms).unwrap();
+            let zigbee_effect_duration = (zigbee_data & 0xff) as u8; // last byte is effect duration
+            assert_eq!(
+                ed.0, zigbee_effect_duration,
+                "Failed to convert {input_ms}ms ({nearest_second}s) into effect duration {zigbee_effect_duration}"
+            );
+        }
```
[the remark this answers is not in the document]
Comment (duvholt) [reply crates/hue/src/effect_duration.rs:194]: I agree, but at the same time I don't see any clean way to map all possible values 1ms to 6h to set of four intervals (1s, 15s, 1min, 5min) without these edge cases. 
I don't have much time this weekend to discuess this more, but here's a little sneak peek on the calculation I did in a spreadsheet:
image

First column is input (in seconds), 2nd is calculation using 1s, 3rd using 15s etc. Last column is the value I captured using wireshark. 

So to me the edge cases make sense (they are always between two intervals) as a way brute force 1ms-6h into a byte, but I hope there is a cleaner mathematical function for this 😄

[the remark this answers is not in the document]
Comment (chrivers) [reply crates/hue/src/effect_duration.rs:194]: As we talked about on discord, I sat down and took a long, hard look at this :sweat_smile: 

After gluing together a *whole* bunch of scripts, I came up with something that could automatically brute-force values from a hue bridge, by sniffing the zigbee packets.

Long story short, I came up with this approximation, that covers the entire valid input range:

```python
        s = math.sqrt(x)
        if s < 12.9842:
            d = 252.003089171134 - 4.61710111004978 * s
        elif s < 62.8835:
            d = 204.137662520171 - 0.930652013572797 * s
        elif s < 188.7587:
            d = 165.296007572164 - 0.312975901543635 * s
        elif s < 752.8070:
            d = 120.951876550729 - 0.0780508937762312 * s
        elif s < 4635.9465:
            d = 73.8115058857269 - 0.0154314333912542 * s
```

It's so freaking random. I have *no* idea why this model fits, and my gut tells me there must be some simpler version of this, that we just haven't thought of.

Now, this model isn't perfect, but it's *never* more than 1 count off the correct result, across all 3691 samples I've taken. (and 96% of the samples are an exact match!)

So I would say this is "good enough" for now. Maybe some day we can discover exactly what's going on, but at least this model is relatively simple, well-tested, more correct than the current implementation, and quite compact.

Would you like to incorporate it into this PR? It would make sense, since you're working on that part of the code anyway :)

File: crates/hue/src/api/light.rs (hunk 1 above)
Comment (chrivers) [reply crates/hue/src/api/light.rs:504]: That's fine :+1: 

It also answers another question I had in the back of my mind: Should LightEffect and EffectType (from the zigbee impl) be merged, since they used to be identical.

The answer now is clearly: no. :)

File: crates/hue/src/effect_duration.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrivers) [reply crates/hue/src/effect_duration.rs:194]: Okay okay, forget that junk I said a long time ago... yesterday :sweat_smile: 

I finally fiddled around with it, until it matches *perfectly*:

```rust
fn convert(x: u32) -> u8 {
    match x {
        _ if x < 1500 => 251,
        _ if x < (100 * 600) => (250 - (x - 1500) / 1000) as u8,
        _ if x < (100 * 2880) => (192 - (x - 57500) / 5000) as u8,
        _ if x < (100 * 2945) => (192 - (x - 64000) / 5000) as u8,
        _ if x < (100 * 8840) => (145 - (x - 307499) / 15000) as u8,
        _ if x < (100 * 8845) => 107,
        _ if x < (100 * 34495) => (106 - (x - 929499) / 60000) as u8,
        _ if x < (100 * 35395) => 63,
        _ if x < (100 * 37495) => 62,
        _ if x < (100 * 214495) => (61 - (x - 4049499) / 300000) as u8,
        _ => 2,
    }
}
```

This one gets every (new) sample correct, with no errors, and no under/overflows on integer math.

[the remark this answers is not in the document]
Comment (chrivers) [reply crates/hue/src/effect_duration.rs:194]: So, to follow up on the discord messages, I put our implementations together, and tried to get the best of both world. The advantages include:

1) Code is easier to follow (from your effort)
2) Code is free of floating point math (from my attempt)
3) They are now `const fn` (as a side effect of 2)

So this is the new *entire* impl for EffectDuration:
```rust

impl EffectDuration {
    const RESOLUTION_01S_BASE: u8 = 0xFC;
    const RESOLUTION_05S_BASE: u8 = 0xCC;
    const RESOLUTION_15S_BASE: u8 = 0xA5;
    const RESOLUTION_01M_BASE: u8 = 0x79;
    const RESOLUTION_05M_BASE: u8 = 0x4A;

    const RESOLUTION_01S: u32 = 1; // 1s.
    const RESOLUTION_05S: u32 = 5; // 5s.
    const RESOLUTION_15S: u32 = 15; // 15s.
    const RESOLUTION_01M: u32 = 60; // 1min.
    const RESOLUTION_05M: u32 = 5 * 60; // 5min.

    pub const fn from_ms(milliseconds: u32) -> HueResult<Self> {
        Self::from_seconds(milliseconds / 1000)
    }

    #[allow(clippy::cast_possible_truncation)]
    pub const fn from_seconds(seconds: u32) -> HueResult<Self> {
        let num = match seconds {
            0..60 => Self::RESOLUTION_01S_BASE - (seconds / Self::RESOLUTION_01S) as u8,
            // ~5min
            60..293 => Self::RESOLUTION_05S_BASE - (seconds / Self::RESOLUTION_05S) as u8,
            // 293 and 294 do not fit into any of the bases as they both output 146
            293..=294 => 146,
            // ~15min
            295..878 => Self::RESOLUTION_15S_BASE - (seconds / Self::RESOLUTION_15S) as u8,
            878..885 => 107,
            // ~60min
            885..3510 => Self::RESOLUTION_01M_BASE - (seconds / Self::RESOLUTION_01M) as u8,
            3510..3540 => 63,
            // 06hrs
            3540..=21600 => Self::RESOLUTION_05M_BASE - (seconds / Self::RESOLUTION_05M) as u8,

            _ => return Err(HueError::EffectDurationOutOfRange),
        };

        Ok(Self(num))
    }
}
```

This has been tested on all 21600000 possible inputs, and gives the correct result in every case. I think this is a good candidate for the version we're going to use. Comments?

I'll follow up with a sketch for unit testing :+1:

File: src/backend/z2m/backend_event.rs [hunk 3]
```diff
@@ -63,6 +64,17 @@ impl Z2mBackend {
...
         }
 
+        if let Some(act) = &upd.timed_effects {
+            if let Some(fx) = act.effect {
+                hz = hz.with_effect_type(fx.into());
+            }
+
+            if let Some(duration) = act.duration {
+                let EffectDuration(effect_duration) = EffectDuration::from_ms(duration)?;
+                hz = hz.with_effect_speed(effect_duration);
```
[the remark this answers is not in the document]
Comment (chrivers) [reply src/backend/z2m/backend_event.rs:74]: That way we're actually gaining something from the newtype pattern :+1:

File: crates/hue/src/effect_duration.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrivers) [reply crates/hue/src/effect_duration.rs:194]: For unit testing, I think it's worthwhile to include to complete table of breakpoints (inside `mod tests {}`):

```rust
#[allow(clippy::unreadable_literal)]
const DURATION_BREAKPOINTS: &[(u32, u8)] = &[
    (1499, 251),
    (2499, 250),
    (3499, 249),
    (4499, 248),
    (5499, 247),
    (6499, 246),
    (7499, 245),
    (8499, 244),
    (9499, 243),
    (10499, 242),
    (11499, 241),
    (12499, 240),
    (13499, 239),
    (14499, 238),
    (15499, 237),
    (16499, 236),
    (17499, 235),
    (18499, 234),
    (19499, 233),
    (20499, 232),
    (21499, 231),
    (22499, 230),
    (23499, 229),
    (24499, 228),
    (25499, 227),
    (26499, 226),
    (27499, 225),
    (28499, 224),
    (29499, 223),
    (30499, 222),
    (31499, 221),
    (32499, 220),
    (33499, 219),
    (34499, 218),
    (35499, 217),
    (36499, 216),
    (37499, 215),
    (38499, 214),
    (39499, 213),
    (40499, 212),
    (41499, 211),
    (42499, 210),
    (43499, 209),
    (44499, 208),
    (45499, 207),
    (46499, 206),
    (47499, 205),
    (48499, 204),
    (49499, 203),
    (50499, 202),
    (51499, 201),
    (52499, 200),
    (53499, 199),
    (54499, 198),
    (55499, 197),
    (56499, 196),
    (57499, 195),
    (58499, 194),
    (59499, 193),
    (62499, 192),
    (67499, 191),
    (72499, 190),
    (77499, 189),
    (82499, 188),
    (87499, 187),
    (92499, 186),
    (97499, 185),
    (102499, 184),
    (107499, 183),
    (112499, 182),
    (117499, 181),
    (122499, 180),
    (127499, 179),
    (132499, 178),
    (137499, 177),
    (142499, 176),
    (147499, 175),
    (152499, 174),
    (157499, 173),
    (162499, 172),
    (167499, 171),
    (172499, 170),
    (177499, 169),
    (182499, 168),
    (187499, 167),
    (192499, 166),
    (197499, 165),
    (202499, 164),
    (207499, 163),
    (212499, 162),
    (217499, 161),
    (222499, 160),
    (227499, 159),
    (232499, 158),
    (237499, 157),
    (242499, 156),
    (247499, 155),
    (252499, 154),
    (257499, 153),
    (262499, 152),
    (267499, 151),
    (272499, 150),
    (277499, 149),
    (282499, 148),
    (287499, 147),
    (294499, 146),
    (307499, 145),
    (322499, 144),
    (337499, 143),
    (352499, 142),
    (367499, 141),
    (382499, 140),
    (397499, 139),
    (412499, 138),
    (427499, 137),
    (442499, 136),
    (457499, 135),
    (472499, 134),
    (487499, 133),
    (502499, 132),
    (517499, 131),
    (532499, 130),
    (547499, 129),
    (562499, 128),
    (577499, 127),
    (592499, 126),
    (607499, 125),
    (622499, 124),
    (637499, 123),
    (652499, 122),
    (667499, 121),
    (682499, 120),
    (697499, 119),
    (712499, 118),
    (727499, 117),
    (742499, 116),
    (757499, 115),
    (772499, 114),
    (787499, 113),
    (802499, 112),
    (817499, 111),
    (832499, 110),
    (847499, 109),
    (862499, 108),
    (884499, 107),
    (929499, 106),
    (989499, 105),
    (1049499, 104),
    (1109499, 103),
    (1169499, 102),
    (1229499, 101),
    (1289499, 100),
    (1349499, 99),
    (1409499, 98),
    (1469499, 97),
    (1529499, 96),
    (1589499, 95),
    (1649499, 94),
    (1709499, 93),
    (1769499, 92),
    (1829499, 91),
    (1889499, 90),
    (1949499, 89),
    (2009499, 88),
    (2069499, 87),
    (2129499, 86),
    (2189499, 85),
    (2249499, 84),
    (2309499, 83),
    (2369499, 82),
    (2429499, 81),
    (2489499, 80),
    (2549499, 79),
    (2609499, 78),
    (2669499, 77),
    (2729499, 76),
    (2789499, 75),
    (2849499, 74),
    (2909499, 73),
    (2969499, 72),
    (3029499, 71),
    (3089499, 70),
    (3149499, 69),
    (3209499, 68),
    (3269499, 67),
    (3329499, 66),
    (3389499, 65),
    (3449499, 64),
    (3539499, 63),
    (3749499, 62),
    (4049499, 61),
    (4349499, 60),
    (4649499, 59),
    (4949499, 58),
    (5249499, 57),
    (5549499, 56),
    (5849499, 55),
    (6149499, 54),
    (6449499, 53),
    (6749499, 52),
    (7049499, 51),
    (7349499, 50),
    (7649499, 49),
    (7949499, 48),
    (8249499, 47),
    (8549499, 46),
    (8849499, 45),
    (9149499, 44),
    (9449499, 43),
    (9749499, 42),
    (10049499, 41),
    (10349499, 40),
    (10649499, 39),
    (10949499, 38),
    (11249499, 37),
    (11549499, 36),
    (11849499, 35),
    (12149499, 34),
    (12449499, 33),
    (12749499, 32),
    (13049499, 31),
    (13349499, 30),
    (13649499, 29),
    (13949499, 28),
    (14249499, 27),
    (14549499, 26),
    (14849499, 25),
    (15149499, 24),
    (15449499, 23),
    (15749499, 22),
    (16049499, 21),
    (16349499, 20),
    (16649499, 19),
    (16949499, 18),
    (17249499, 17),
    (17549499, 16),
    (17849499, 15),
    (18149499, 14),
    (18449499, 13),
    (18749499, 12),
    (19049499, 11),
    (19349499, 10),
    (19649499, 9),
    (19949499, 8),
    (20249499, 7),
    (20549499, 6),
    (20849499, 5),
    (21149499, 4),
    (21449499, 3),
];
```

Then, a complete conformance test looks like this:

```rust
    let mut c = 1;
    for (x, y) in DURATION_BREAKPOINTS {
        while c < *x {
            assert_eq!(EffectDuration::from_ms(*x), y);
            c += 1;
        }
    }
```

This takes practically no time to run, so it's completely fine to include as a standard test case.

[the remark this answers is not in the document]
Comment (chrivers) [reply crates/hue/src/effect_duration.rs:194]: Also, in this version of the `from_seconds()` impl, I don't know if the comments (`// ~5min`, `// ~15min`, etc) really add all that much, since the main logic is now more of a table, which is reasonably easy to follow.

It's completely up to you - keep them if you like, or remove them, as you prefer :+1:

**Comment by chrivers** [comment]
Fantastic work :+1:

Let's get it merged :)
